cmd/faucet/faucet.go

```diff
@@ -178,288 +134,164 @@ func validAddress(addr string) error {
-			http.Error(w, "rate limited", http.StatusTooManyRequests)
-			return
-		}
-		if err != nil {
-			log.Printf("error sending to %s: %s", req.Address, err)
-			http.Error(w, fmt.Sprintf("error executing cmd: %v", err), http.StatusInternalServerError)
-			return
-		}
-		w.Header().Set("Access-Control-Allow-Origin", "*")
-		w.Header().Set("Content-Type", "application/json")
-		w.WriteHeader(http.StatusOK)
-		if _, err := w.Write([]byte("request batched\n")); err != nil {
-			http.Error(w, fmt.Sprintf("error writing response: %v", err), http.StatusInternalServerError)
-			return
-		}
+	out, err := execute(cmd)
+	if err != nil {
+		return "", http.StatusInternalServerError, err
 	}
-}
 
-func getIP(r *http.Request) (string, error) {
-	forwardedFor := r.Header.Get("X-Real-Ip")
-	if forwardedFor != "" {
-		hops := strings.SplitN(forwardedFor, ",", 2)
-		client := strings.TrimSpace(hops[0])
-		return client, nil
+	var result struct {
+		Code   int    `json:"code"`
+		TxHash string `json:"txhash"`
 	}
 
-	ip, _, err := net.SplitHostPort(r.RemoteAddr)
-	if err != nil {
-		return "", fmt.Errorf("userip: %q is not IP:port", r.RemoteAddr)
+	if err = json.Unmarshal(out.Stdout, &result); err != nil {
+		return "", http.StatusInternalServerError, fmt.Errorf("error unmarshalling tx result: %w", err)
 	}
-	return ip, nil
-}
 
-func envOrDefault(key, defaultValue string) string {
-	v := os.Getenv(key)
-	if v == "" {
-		return defaultValue
+	if result.Code != 0 {
+		return "", http.StatusInternalServerError, fmt.Errorf(
+			"tx failed with code %d for address %s",
+			result.Code,
+			addr,
+		)
 	}
-	return v
-}
 
-type Limiter struct {
-	cooldown time.Duration
-	last     map[string]time.Time
-	mu       sync.Mutex
-}
+	f.TokensAvailable = (f.TokensAvailable - float64(len(f.Batch))*f.Amount/uwardConversion)
+	f.log.Info().Msgf("tokens available: %f", f.TokensAvailable)
 
-func NewLimiter(cooldown time.Duration) *Limiter {
-	return &Limiter{
-		cooldown: cooldown,
-		last:     make(map[string]time.Time),
-	}
+	f.Batch = []string{}
+	return result.TxHash, http.StatusOK, nil
 }
 
-func (l *Limiter) Allow(keys ...string) (func(), error) {
-	l.mu.Lock()
-	defer l.mu.Unlock()
-	for _, k := range keys {
-		if time.Since(l.last[k]) < l.cooldown {
-			return nil, fmt.Errorf("%w: key '%s' must wait %s", ErrRateLimited, k, time.Until(l.last[k].Add(l.cooldown)).String())
-		}
-	}
-	for _, k := range keys {
-		l.last[k] = time.Now()
-	}
-	return func() {
-		l.Reset(keys...)
-	}, nil
-}
+// DailyRefresh.
+func (f *Faucet) DailyRefresh() {
+	for {
+		now := time.Now()
+		nextDay := now.AddDate(0, 0, 1).Truncate(dailyHours * time.Hour)
+		durationUntilNextDay := time.Until(nextDay)
+
+		f.log.Info().Msgf("next token refresh in %s", durationUntilNextDay)
+		time.Sleep(durationUntilNextDay)
 
-func (l *Limiter) Reset(key ...string) {
-	l.mu.Lock()
-	defer l.mu.Unlock()
-	for _, k := range key {
-		l.last[k] = time.Time{}
+		f.TokensAvailable = f.DailySupply
```

Comment:
you should acquire the lock otherwise there is a race condition between this and f.Send() trying to read/write this var

Reply:
Added lock